File: python/leetcode/leetcode19.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        nodes = []
        cur = head

        while cur:
            nodes.append(cur)
            cur = cur.next

        remove_idx = len(nodes) - n
        if remove_idx == 0:
            return head.next

        nodes[remove_idx - 1].next = nodes[remove_idx].next

        return head

if __name__ == "__main__":
    s = Solution()
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    print(s.removeNthFromEnd(head, 2))

File: java/src/leetcode19.java
import java.util.ArrayList;
import java.util.List;

/**
 * @author laege
 * @version 1.0
 * @since 2026-01-20
 */

class Leetcode19 {
    class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        List<ListNode> list = new ArrayList<>();
        ListNode cur = head;
        while (cur != null) {
            list.add(cur);
            cur = cur.next;
        }

        int idx = list.size() - n;

        if (idx == 0) {
            return head.next;
        }

        list.get(idx - 1).next = list.get(idx).next;

        return head;
    }
}
